File: Back_Tracking/m-Coloring/src.py
def promising(i: int) -> bool:
    j = 1
    while j < i:
        if W[i][j] and vcolor[i] == vcolor[j]:
            return False
        j += 1
    return True


def m_coloring(i: int, m: int) -> bool:
    global flag
    global cnt
    if promising(i):
        if i == ver_num:
            flag = True
            cnt += 1
        else:
            for color in range(1, m + 1):
                vcolor[i + 1] = color
                m_coloring(i + 1, m)


if __name__ == "__main__":
    ver_num, edge_num = map(int, input().split())
    W = [[0] * (ver_num + 1) for _ in range(ver_num + 1)]
    for _ in range(edge_num):
        i, j = map(int, input().split())
        W[i][j] = 1
        W[j][i] = 1

    # print(*W, sep='\n')

    vcolor = [0] * (ver_num + 1)

    for m in range(1, ver_num + 1):
        flag = False
        cnt = 0
        m_coloring(0, m)
        if flag:
            print(m)
            print(cnt)
            break

File: Back_Tracking/m-Coloring/src.cpp
#include <iostream>
#include <vector>


using namespace std;


int ver_num, edge_num, cnt, m;
vector<vector<int>> W;
vector<int> vcolor;


bool promising(int i){
    int j = 1;
    while (j < i){
        if (W[i][j] && (vcolor[i] == vcolor[j]))
            return false;
        j++;
    }
    return true;
}


void m_coloring(int i){
    if (promising(i)){
        if (i == ver_num)
            cnt++;
        else{
            for(int color = 1; color <= m; color++){
                vcolor[i + 1] = color;
                m_coloring(i + 1);
            }
        }
    }
}


int main(){
    cin >> ver_num >> edge_num;

    W.resize(ver_num + 1, vector<int>(ver_num + 1, 0));

    int i, j;
    for(int k = 0; k < edge_num; k++){
        cin >> i >> j;
        W[i][j] = 1;
        W[j][i] = 1;
    }

    vcolor.resize(ver_num + 1, 0);

    for(m = 1; m <= ver_num; m++){
        cnt = 0;
        m_coloring(0);
        if (cnt)
            break;
    }

    cout << m << '\n' << cnt << '\n';

    return 0;
}
